### PR TITLE
Limit the use of the same playlist name when adding a playlist

### PR DESCRIPTION
**Pull request by AmbrishRamachandiran**
## Hey, I just made a Pull Request!

When user try to add the playlist its should not accept the duplicate playlist name , duplicates which will leads to confusions and having an unique playlist name is the good fast approach to find the saved playlist from the list.

Here the implementation screens :

image

image

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by AmbrishRamachandiran** [comment]
@kuangp Please review the above PR ! the current behavior of playlist creation/edit accepting the duplicate playlist name which is not a common approach ! 
i have made the changes to prevent/limit the duplicate playlist name/title please let me know if any further changes are required

File: plugins/playlist/src/components/PlaylistPage/PlaylistHeader.tsx
Comment (freben) [inline plugins/playlist/src/components/PlaylistPage/PlaylistHeader.tsx:86]: Don't put this in a state. Just call `alertApi.post` whenever you want, directly in code. Otherwise you'll unnecessarily re-render this component twice (once when setting the message and once after clearing it again)

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (freben) [inline plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:79]: Similar to the other pull request you made, you should probably re-request this list when you do the verification, not just once. Don't ever do significant work like this in the actual render method of components ; this belongs in a `useEffect` or `useCallback` or similar. That is a general React rule overall.

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (freben) [inline plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:83]: Not really processing anything, this function should be removed.

To be clear, calling `await playListApiData` here only waits for the one, single, request that was made above. It never re-fetches anything.

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (freben) [inline plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:109]: Wait, is the code in this if statement and the one above exactly the same, or am I seeing it wrong?

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (freben) [inline plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:103]: hm this is a bit of an odd way of expressing it, should probably be `.some(p => p.name === name)`

**Comment by AmbrishRamachandiran** [comment]
@freben requested changes are done please review and let me know any other changes are required 😊

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (freben) [inline plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:89]: This still just does it once. Oh well the dialog is short lived so it's probably fine.

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (freben) [inline plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:103]: Now there's a race condition instead, such that if the playlist hasn't loaded yet it can't validate properly (it'll just see the empty array). What I suggested was to call `getAllPlaylists` INSIDE here, without a `useState`.

But alright you know what, if we want to keep the just-do-one-request pattern, you can just do `const playlistPromise = useRef(playlistApi.getAllPlaylists({ editable: true }));` and delete the `useEffect`. Then it's is a `Promise<PlaylistMetadata[]>`, so your `nameIsUnique` function can do `const playlist = await playlistPromise.current;` to get the actual value. That resolves the race

**Review by freben — COMMENTED** [review]
Alright back over to @kuangp

**Comment by AmbrishRamachandiran** [comment]
@freben added fixes as per you suggestion please validate for approval

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (kuangp) [inline plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:80]: We probably don't want to limit this to `editable` playlists - this would only include playlists the user has permissions to modify and not every playlist the user has permissions to `read` so duplicates could still appear in the list of every playlists visible to the user 
```suggestion
    playlistApi.getAllPlaylists(),
```

**Review by kuangp — CHANGES_REQUESTED** [review]

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (AmbrishRamachandiran) [reply plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx]: Typescript build failed 
after  playlistApi.getAllPlaylists(), so passed empty object to resolve is 
 playlistApi.getAllPlaylists({}),

**Comment by AmbrishRamachandiran** [comment]
@kuangp Added changes as per your suggestion please review.

File: plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx
Comment (kuangp) [inline plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:87]: I'm still not sure why we need the `isEditing` param?

Comment (AmbrishRamachandiran) [reply plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:87]: It is for when user tries to edit a playlist and he just updated the description or other field except title field its should not throw the duplicate error

Comment (kuangp) [reply plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:87]: Isn't that already satisfied by the `name !== playlist.name` check? If someone is not changing the name then this conditional should fail since the name has not changed right? and it should fall back to the `return true;` line below which indicates there is no error? 

Let me know if I'm missing something

Comment (AmbrishRamachandiran) [reply plugins/playlist/src/components/PlaylistEditDialog/PlaylistEditDialog.tsx:87]: > I'm still not sure why we need the `isEditing` param? 

When user tries to edit the playlist and if he start enters some value in the title field the validation will be started or else its should not throw any errors 

If user wants to edit just a description we need to allow them to edit it without any validations.

Will try to remove the isEdit param and test it once and get back to you.

**Comment by AmbrishRamachandiran** [comment]
@camilaibs Waiting for your approval please review TIA.

**Comment by AmbrishRamachandiran** [comment]
@freben @Rugvip @benjdlambert @camilaibs any chance to merge this today?

**Comment by AmbrishRamachandiran** [comment]
@freben @Rugvip @benjdlambert can i get merge approval today ?

**Comment by AmbrishRamachandiran** [comment]
@freben @benjdlambert @Rugvip awaiting merge😊

**Comment by sennyeya** [comment]
I think this is causing an error message for guest users, see #19794.

**Comment by AmbrishRamachandiran** [comment]
@sennyeya we can't use playlist page without login into backstage , the page is developed like that only & this feature is not related to your issue.

Similar to that we also cant use backstage entity component without login.

**Comment by sennyeya** [comment]
@AmbrishRamachandiran This feature is definitely related to my bug, we started preloading the list of playlists in PlaylistEditDialog. That preloading has no error handling so if I open the page as a guest user or unauthorised user, an error boundary is shown with the message described in my bug ticket. 

I'll leave it up to the playlist plug-in maintainers how they'd like the UX to be, but I'd like to use guest mode :)

**Pull request merged**
